Userland/Libraries/LibCore/EventLoop.cpp

```diff
@@ -266,40 +271,44 @@ EventLoop::EventLoop([[maybe_unused]] MakeInspectable make_inspectable)
         s_timers = new HashMap<int, NonnullOwnPtr<EventLoopTimer>>;
         s_notifiers = new HashTable<Notifier*>;
     }
-
-    if (!s_main_event_loop) {
-        s_main_event_loop = this;
-        s_pid = getpid();
+    s_main_event_loop.with_locked([&, this](auto*& main_event_loop) {
+        if (main_event_loop == nullptr) {
+            // FIXME: The compiler complains that we don't use main_event_loop although we set it.
+            main_event_loop = this;
+            s_pid = getpid();
+            // FIXME: We only create the wake pipe for the main thread
 #if defined(SOCK_NONBLOCK)
-        int rc = pipe2(s_wake_pipe_fds, O_CLOEXEC);
+            int rc = pipe2(s_wake_pipe_fds, O_CLOEXEC);
 #else
-        int rc = pipe(s_wake_pipe_fds);
-        fcntl(s_wake_pipe_fds[0], F_SETFD, FD_CLOEXEC);
-        fcntl(s_wake_pipe_fds[1], F_SETFD, FD_CLOEXEC);
+            int rc = pipe(s_wake_pipe_fds);
+            fcntl(s_wake_pipe_fds[0], F_SETFD, FD_CLOEXEC);
+            fcntl(s_wake_pipe_fds[1], F_SETFD, FD_CLOEXEC);
 
 #endif
-        VERIFY(rc == 0);
-        s_event_loop_stack->append(*this);
+            VERIFY(rc == 0);
+            s_event_loop_stack->append(*this);
 
 #ifdef __serenity__
-        if (getuid() != 0
-            && make_inspectable == MakeInspectable::Yes
-            && !s_inspector_server_connection) {
-            if (!connect_to_inspector_server())
-                dbgln("Core::EventLoop: Failed to connect to InspectorServer");
-        }
+            if (getuid() != 0
+                && make_inspectable == MakeInspectable::Yes
+                // FIXME: Deadlock potential; though the main loop and inspector server connection are rarely used in conjunction
+                && s_inspector_server_connection.with_locked([](auto inspector_server_connection) { return inspector_server_connection; })) {
+                if (!connect_to_inspector_server())
+                    dbgln("Core::EventLoop: Failed to connect to InspectorServer");
+            }
 #endif
-    }
+        }
+    });
 
     dbgln_if(EVENTLOOP_DEBUG, "{} Core::EventLoop constructed :)", getpid());
 }
 
 EventLoop::~EventLoop()
 {
     // NOTE: Pop the main event loop off of the stack when destroyed.
-    if (this == s_main_event_loop) {
+    if (s_main_event_loop.with_locked([this](auto* main_event_loop) { return this == main_event_loop; })) {
         s_event_loop_stack->take_last();
-        s_main_event_loop = nullptr;
+        s_main_event_loop.with_locked([]([[maybe_unused]] auto*& main_event_loop) { main_event_loop = nullptr; });
```

Comment:
Is there a TOCTOU here? Should we keep the `s_main_event_loop` locked the entire time?

Reply:
Absolutely, I already discussed this with Andrew so it's prime time to do it properly.